tools/cli/workflowCommands.go

```diff
@@ -938,34 +938,53 @@ func convertDescribeWorkflowExecutionResponse(resp *workflowservice.DescribeWork
 func convertSearchAttributes(searchAttributes *commonpb.SearchAttributes,
 	wfClient workflowservice.WorkflowServiceClient, c *cli.Context) *clispb.SearchAttributes {
 
-	if searchAttributes == nil || len(searchAttributes.GetIndexedFields()) == 0 {
+	if len(searchAttributes.GetIndexedFields()) == 0 {
 		return nil
 	}
 
 	result := &clispb.SearchAttributes{
 		IndexedFields: map[string]string{},
 	}
-	ctx, cancel := newContext(c)
-	defer cancel()
-	validSearchAttributes, err := wfClient.GetSearchAttributes(ctx, &workflowservice.GetSearchAttributesRequest{})
-	if err != nil {
-		ErrorAndExit("Error when get search attributes", err)
-	}
-	validKeys := validSearchAttributes.GetKeys()
 
-	indexedFields := searchAttributes.GetIndexedFields()
-	for k, v := range indexedFields {
-		valueType := validKeys[k]
-		deserializedValue, err := common.DeserializeSearchAttributeValue(v, valueType)
+	setSearchAttributesType(searchAttributes, wfClient, c)
+
+	for k, v := range searchAttributes.GetIndexedFields() {
+		deserializedValue, err := searchattribute.Decode(v)
 		if err != nil {
-			ErrorAndExit("Error deserializing search attribute value", err)
+			// In case of error just use raw data (which is JSON).
+			deserializedValue = string(v.GetData())
 		}
 		result.IndexedFields[k] = fmt.Sprintf("%v", deserializedValue)
 	}
 
 	return result
 }
 
+func setSearchAttributesType(searchAttributes *commonpb.SearchAttributes, wfClient workflowservice.WorkflowServiceClient, c *cli.Context) {
+	allTypesAreSet := true
+	for _, p := range searchAttributes.GetIndexedFields() {
+		if _, isTypeSet := p.Metadata[searchattribute.MetadataType]; !isTypeSet {
+			allTypesAreSet = false
+			break
+		}
+	}
+
+	if allTypesAreSet {
+		return
+	}
+
+	// For all newly created search attributes type should be set in Metadata field.
+	// Call to GetSearchAttributes is just for backward compatibility.
+	// It may fail with Unauthorized error which will be safely ignored.
+	ctx, cancel := newContext(c)
+	defer cancel()
+	validSearchAttributes, err := wfClient.GetSearchAttributes(ctx, &workflowservice.GetSearchAttributesRequest{})
+	if err != nil {
+		validSearchAttributes = nil
+	}
+	searchattribute.SetType(searchAttributes, validSearchAttributes.GetKeys())
```

Comment:
Wouldn't `validSearchAttributes = nil` cause an NRE here?

Reply:
This is `go` magic. All `GetXXX()` autogenerated methods check receiver for `nil`. I can refactor if it looks confusing/scared.